lib/src/kvs/ds.rs

```diff
@@ -251,24 +275,201 @@ impl Datastore {
+				}
+			}
+		}
+		Ok(archived)
+	}
+
+	pub async fn remove_archived(
+		&self,
+		tx: &mut Transaction,
+		archived: Vec<LqValue>,
+	) -> Result<(), Error> {
+		for lq in archived {
+			// Delete the cluster key, used for finding LQ associated with a node
+			tx.del(lq::new(lq.cl, lq.ns.as_str(), lq.db.as_str(), lq.lq)).await?;
+			// Delete the table key, used for finding LQ associated with a table
+			tx.del(Lv::new(lq.ns.as_str(), lq.db.as_str(), lq.tb.as_str(), lq.lq)).await?;
+		}
+		Ok(())
+	}
+
+	pub async fn _garbage_collect(
+		// TODO not invoked
+		// But this is garbage collection outside of bootstrap
+		&self,
+		tx: &mut Transaction,
+		watermark: &Timestamp,
+		this_node_id: &Uuid,
+	) -> Result<(), Error> {
+		let dead_heartbeats = self.delete_dead_heartbeats(tx, watermark).await?;
+		trace!("Found dead hbs: {:?}", dead_heartbeats);
+		let mut archived: Vec<LqValue> = vec![];
+		for hb in dead_heartbeats {
+			let new_archived = self.archive_lv_for_node(tx, &hb.nd, this_node_id).await?;
+			tx.del_cl(hb.nd).await?;
+			trace!("Deleted node {}", hb.nd);
+			for lq_value in new_archived {
+				archived.push(lq_value);
+			}
+		}
 		Ok(())
 	}
 
+	// Returns a list of live query IDs
+	pub async fn archive_lv_for_node(
+		&self,
+		tx: &mut Transaction,
+		nd: &Uuid,
+		this_node_id: &Uuid,
+	) -> Result<Vec<LqValue>, Error> {
+		let lqs = tx.all_lq(nd).await?;
+		trace!("Archiving lqs and found {} LQ entries for {}", lqs.len(), nd);
+		let mut ret = vec![];
+		for lq in lqs {
+			let lvs = tx.get_lv(lq.ns.as_str(), lq.db.as_str(), lq.tb.as_str(), &lq.lq).await?;
+			let archived_lvs = lvs.clone().archive(*this_node_id);
+			tx.putc_lv(&lq.ns, &lq.db, &lq.tb, archived_lvs, Some(lvs)).await?;
+			ret.push(lq);
+		}
+		Ok(ret)
+	}
+
+	/// Given a timestamp, delete all the heartbeats that have expired
+	/// Return the removed heartbeats as they will contain node information
+	pub async fn delete_dead_heartbeats(
+		&self,
+		tx: &mut Transaction,
+		ts: &Timestamp,
+	) -> Result<Vec<Hb>, Error> {
+		let limit = 1000;
+		let dead = tx.scan_hb(ts, limit).await?;
+		tx.delr_hb(dead.clone(), 1000).await?;
+		for dead_node in dead.clone() {
+			tx.del_cl(dead_node.nd).await?;
+		}
+		Ok::<Vec<Hb>, Error>(dead)
+	}
+
 	// Creates a heartbeat entry for the member indicating to the cluster
-	// that the node is alive
+	// that the node is alive.
 	pub async fn heartbeat(&self) -> Result<(), Error> {
 		let mut tx = self.transaction(true, false).await?;
-		tx.set_hb(self.id).await?;
-		tx.commit().await?;
-		Ok(())
+		let timestamp = tx.clock();
+		self.heartbeat_full(&mut tx, timestamp, self.id).await?;
+		tx.commit().await
 	}
 
+	// Creates a heartbeat entry for the member indicating to the cluster
+	// that the node is alive. Intended for testing.
+	pub async fn heartbeat_full(
+		&self,
+		tx: &mut Transaction,
+		timestamp: Timestamp,
+		node_id: Uuid,
+	) -> Result<(), Error> {
+		tx.set_hb(timestamp, node_id).await
+	}
```

Comment:
We will probably want this down the line. For scenarios where clusters have no sent recent heartbeats. We arent testing that at the moment. Otherwise, `heartbeat()` uses the available clock, which would otherwise need to be mocked.

Reply:
Can this be made clearer in the comments? I'm not sure how this function is different / is going to be different to the `heartbeat()` function except for the `Intended for testing.` comment, but I don't see tests which use this function.

Reply:
In the codebase we would use heartbeat, because all the parameters that are hard to control (time, filesystems, networks, etc) would be centralised (`clock.now()`, `obvious_file.read()`, `the_one_and_only_connection.read()`). But by allowing them to be injected, we can carefully control how they are used for tests
```
fn test_time_shift_gets_detected() {
   clock = fake_clock(["12:00", "12:01", "11:02"])
   ds = fake_ds()
   output = function_we_are_testing_such_as_heartbeats(ctx, clock, ds)
   assert_eq!(output, expected)
   assert_true!(some_property(ds))
}
```

Reply:
Another example is race conditions
```
fn test_conflicts_are_retried() {
   kv = fake_kv_errors_first(
      errors: [TX_ISOLATION, TX_ISOLATION]
      results: ["works"]
   )
   assert_error(tested_function(kv, retries: 2)
   assert_ok(tested_function(kv, retries: 3)
}
```

Reply:
Added more comments but kept function